hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/PipelineStateManagerV2.java

```diff
@@ -0,0 +1,101 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.hadoop.hdds.scm.pipeline;
+
+import org.apache.hadoop.hdds.protocol.DatanodeDetails;
+import org.apache.hadoop.hdds.protocol.proto.HddsProtos;
+import org.apache.hadoop.hdds.scm.container.ContainerID;
+import org.apache.hadoop.hdds.scm.metadata.Replicate;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.NavigableSet;
+
+/**
+ * Manages the state of pipelines in SCM.
+ */
+public interface PipelineStateManagerV2 {
+
+  /**
+   * Adding pipeline would be replicated to Ratis.
+   * @param pipelineProto
+   * @throws IOException
+   */
+  @Replicate
+  void addPipeline(HddsProtos.Pipeline pipelineProto) throws IOException;
+
+  /**
+   * Removing pipeline would be replicated to Ratis.
+   * @param pipelineIDProto
+   * @return Pipeline removed
+   * @throws IOException
+   */
+  @Replicate
+  void removePipeline(HddsProtos.PipelineID pipelineIDProto)
+      throws IOException;
+
+  /**
+   * Updating pipeline state would be replicated to Ratis.
+   * @param pipelineIDProto
+   * @param newState
+   * @throws IOException
+   */
+  @Replicate
+  void updatePipelineState(HddsProtos.PipelineID pipelineIDProto,
+                           HddsProtos.PipelineState newState)
+      throws IOException;
+
+  void addContainerToPipeline(PipelineID pipelineID,
+                              ContainerID containerID) throws IOException;
+
+  Pipeline getPipeline(PipelineID pipelineID) throws PipelineNotFoundException;
+
+  List<Pipeline> getPipelines();
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              HddsProtos.ReplicationFactor factor);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              HddsProtos.ReplicationFactor factor,
+                              Pipeline.PipelineState state);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              HddsProtos.ReplicationFactor factor,
+                              Pipeline.PipelineState state,
+                              Collection<DatanodeDetails> excludeDns,
+                              Collection<PipelineID> excludePipelines);
+
+  List<Pipeline> getPipelines(HddsProtos.ReplicationType type,
+                              Pipeline.PipelineState... states);
+
+  NavigableSet<ContainerID> getContainers(PipelineID pipelineID)
+      throws IOException;
+
+  int getNumberOfContainers(PipelineID pipelineID) throws IOException;
+
+
+  void removeContainerFromPipeline(PipelineID pipelineID,
+                                   ContainerID containerID) throws IOException;
+
+  void initialize() throws IOException;
```

Comment:
We don't need a `initialize` method exposed, this can. be done called inside `PipelineStateManagerImpl` constructor.

Reply:
Updated.